### ISSUE TITLE
How to extend EF functionality with custom Extension Methods

### ISSUE THREAD

**Issue by pgrm** [aspnet/EntityFramework#5845]
Related to my question on stackoverflow [*How to implement Select For Update in EF Core*](http://stackoverflow.com/questions/37984312/how-to-implement-select-for-update-in-ef-core), I was wondering if it is possible to create custom extension methods in order to modify the executed SQL statement before execution. 

In my case, I want to let other developers define queries in Linq or Fluent methods, take those select statements and append `FOR UPDATE`, `FOR UPDATE NOWAIT`, or `FOR UPDATE SKIP LOCKED`. Another, and very popular, example is [Entity Framework Extended Library](https://github.com/loresoft/EntityFramework.Extended). A colleague of mine told me, that their approach won't work anymore in EF Core. I didn't have time to investigate it yet and I also hoped for a much simpler solution, something like Linq 2 SQL provides with [DataContext.GetCommand](https://msdn.microsoft.com/en-us/library/system.data.linq.datacontext.getcommand(v=vs.110).aspx) but I'm loosing hope that those solutions exist. 

That's why I'd like to know, if it is somehow possible to reuse the existing handling of IQueryable from EF Core and extend it with custom features or I should go along and try to define and create a bunch of fluent methods myself instead.

Thanks

**Comment by igor-gaichenkov-hetras** [comment]
:+1:

**Comment by rowanmiller** [comment]
@anpete  to provide some guidance on what is possible

**Comment by anpete** [comment]
Interception is built-in to EF Core via our DI system. See https://docs.efproject.net/en/latest/miscellaneous/internals/services.html for details, but the general approach will be to subclass or decorate and existing EF component and register it such that EF will use the custom version.

Now, the question becomes, where is the best place to inject specific custom logic? For query, there are several candidates and which is best depends on what you are trying to achieve.

1) Intercept the LINQ expression tree for rewriting. See #4048. 
2) Intercept the relinq QueryModel for rewriting: Register a custom IDatabase impl.
3) Intercept SQL generation via sub-classing the provider's impl. E.g. SqlServerQuerySqlGenerator for SQL Server.

In this case, if you are just looking to modify the generated SQL string directly, then I would go with number 3.

**Comment by pgrm** [comment]
@anpete thank you very much for the hints. I guess I'd know now how to run all select queries with the "FOR UPDATE" suffix. Unfortunately I only want to run few of them. So I would override `VisitSelect(SelectExpression selectExpression)` like this:

````
public override Expression VisitSelect(SelectExpression selectExpression)
{
    base.VisitSelect(selectExpression);
    Sql.Append("FOR UPDATE");
    return selectExpression;
}
````

As I said, `FOR UPDATE` needs to be at the very end, which makes this fairly easy. But I couldn't figure out how I could trigger this. Choosing if I want to run the query with lock or without lock and similar. Initially I thought I could write something like

    from Context.Users.Where(u => u.Name == "Peter").ForUpdate(); 
    // ForUpdate would execute the query and only return IEnumerable<T> / T

But I didn't find a good way of doing that without the need to add a new `*Node` class to Relinq (https://github.com/re-motion/Relinq/tree/master/Core/Parsing/Structure/IntermediateModel).

So basically now I know how I can enhance the SQL - thanks for that. Unfortunately I have no idea how to communicate this need to my custom `QuerySqlGenerator` class. Is there some super easy way I missed? Or should I instead trigger the creation of the query myself, add `FOR UPDATE` and execute it manually?

**Comment by anpete** [comment]
@pgrm Take a look at how we implement the AsNoTracking and Include operators. We are creating custom re-linq result operators that are represented as "query annotations" - Ad hoc pieces of data attached to a query. Once you have the ForUpdate data represented in the query pipeline, it then becomes a matter of injecting somewhere into SQL generation, looking for the annotation (available on QueryCompilationContext) and affecting the generated SelectExpression, possibly via a custom SQL expression node. From that point, as you describe, it just becomes a matter of teaching SQL gen about the new node type.

**Comment by pgrm** [comment]
@anpete thanks, I think I've identified all classes I need to extend / implement but am facing now another issue - I can't override the default classes which are registered in the service.

I've looked at the implementation of [AddEntityFrameworkNpgsql](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/blob/dev/src/Npgsql.EntityFrameworkCore.PostgreSQL/Extensions/NpgsqlServiceCollectionExtensions.cs#L77). So far we didn't need to call that method in our application and dependency injection of our DB context worked anyway. The other class I looked at is [NpgsqlDatabaseProviderServices](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/blob/dev/src/Npgsql.EntityFrameworkCore.PostgreSQL/Storage/Internal/NpgsqlDatabaseProviderServices.cs#L48) which instantiates all the services using the service locator.

This is in my `Startup/ConfigureServices`:

    services.AddScoped<NpgsqlQueryCompilationContextFactory, CustomQueryCompilationContextFactory>();
    services.AddScoped<NpgsqlQuerySqlGeneratorFactory, CustomQuerySqlGeneratorFactory>();
    services.AddEntityFrameworkNpgsql();
    services.AddDbContext<MyDbContext>(options => options.UseNpgsql(connectionString, opt => opt.MigrationsAssembly(currentAssemblyName)));
    services.AddMvc();

When I enumerate `services` I see that it was registered correctly. And if I got it right, than the first registration counts. But when I check a Queryable object the Provider the `NpgsqlQueryCompilationContextFactory` is being used, but I don't understand why, since I thought that

    public override IQuerySqlGeneratorFactory QuerySqlGeneratorFactory => GetService<NpgsqlQuerySqlGeneratorFactory>();

should resolve it to `CustomQuerySqlGeneratorFactory`. I've also tried to register an extended class of `NpgsqlDatabaseProviderServices` and override those methods, but the class also isn't instantiated ever. It seems to me almost as if Entity Framework is using it's own ServiceCollection to set up the classes. What am I doing wrong?

**Comment by anpete** [comment]
@pgrm What does the ctor of your context look like?

**Comment by pgrm** [comment]
@anpete `MyDbContext`? It has 2 constructors.

    public MyDbContext() : base() { }
    public MyDbContext(DbContextOptions<MyDbContext> options) : base(options) { }

**Comment by anpete** [comment]
@pgrm Try calling DbContextOptionsBuilder.UseInternalServiceProvider when configuring options.

**Comment by anpete** [comment]
cc @ajcvickers for how to do this.

**Comment by ajcvickers** [comment]
Something like this:
```C#
services.AddDbContext<MyDbContext>(
    (serviceProvider, options) =>
        options
            .UseInternalServiceProvider(serviceProvider)
            .UseNpgsql(connectionString, opt => opt.MigrationsAssembly(currentAssemblyName)));
```

**Comment by anpete** [comment]
@ajcvickers Thanks!

**Comment by pgrm** [comment]
@ajcvickers and @anpete thank you very much for guiding me throw this - it's working now with that simple example. I tried to extend it as follows:

    public static IQueryable<TEntity> WithLock<TEntity>
        (this IQueryable<TEntity> source, LockStrength lockStrength, string tableName, LockOption? lockOption)
    {
        if (source == null)
        {
            throw new ArgumentNullException(nameof(source));
        }

        return source.Provider.CreateQuery<TEntity>(
            Expression.Call(
                null, 
                WithLockMethodInfo.MakeGenericMethod(typeof(TEntity)), 
                source.Expression,
                Expression.Constant(lockStrength, typeof(LockStrength)), 
                Expression.Constant(tableName, typeof(string)), 
                Expression.Constant(lockOption, typeof(LockOption?))));
    }

to be able to set the lock strength and other options. The inspiration I got from the `FromSqlExpressionNode` class. However, when I define the constructor with `ConstantExpression` parameters like this:

    public class WithLockExpressionNode : ResultOperatorExpressionNodeBase
    {
        public static readonly IReadOnlyCollection<MethodInfo> SupportedMethods = new[]
        {
            Extensions.WithLockMethodInfo
        };

        public WithLockExpressionNode(
            MethodCallExpressionParseInfo parseInfo, 
            ConstantExpression lockStrength, 
            ConstantExpression tableName,
            ConstantExpression lockOption) : base(parseInfo, null, null)
        {
            if (lockStrength.Type != typeof(LockStrength))
            {
                throw new ArgumentException($"Second parameter needs to be of type {typeof(LockStrength)} instead of {lockStrength.Type}");
            }
            if (tableName.Type != typeof(string))
            {
                throw new ArgumentException($"Third parameter needs to be of type {typeof(string)} instead of {tableName.Type}");
            }
            if (lockOption.Type != typeof(LockOption?))
            {
                throw new ArgumentException($"Fourth parameter needs to be of type {typeof(LockOption)} instead of {lockOption.Type}");
            }
        }
    ....

than I get the following exception:

> System.NotSupportedException: Could not parse expression 'value(Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1[NAMESPACE.CLASS]).AsNoTracking().Where(x => (x.Column == __colValue_0)).Select(dto => new DTO_CLASS() {Name = dto.Name, Id = dto.Id}).WithLock(__p_1, __p_2, __p_3)': The given arguments did not match the expected arguments: Object of type 'System.Linq.Expressions.TypedParameterExpression' cannot be converted to type 'System.Linq.Expressions.ConstantExpression'.

When I change the constructor parameters to `ParameterExpression` than it works, but than I couldn't figure out how to retrieve the underlying values from those parameters. Constructor which works:

    public WithLockExpressionNode(
        MethodCallExpressionParseInfo parseInfo, 
        ParameterExpression lockStrength,
        ParameterExpression tableName,
        ParameterExpression lockOption) : base(parseInfo, null, null)
    {
        if (lockStrength.Type != typeof(LockStrength))
        {
            throw new ArgumentException($"Second parameter needs to be of type {typeof(LockStrength)} instead of {lockStrength.Type}");
        }
        if (tableName.Type != typeof(string))
        {
            throw new ArgumentException($"Third parameter needs to be of type {typeof(string)} instead of {tableName.Type}");
        }
        if (lockOption.Type != typeof(LockOption?))
        {
            throw new ArgumentException($"Fourth parameter needs to be of type {typeof(LockOption)} instead of {lockOption.Type}");
        }
    }

I also thought that I might need to access those parameters in my custom `QuerySqlGenerator` but I also couldn't find easy to understand examples how that could work.

Thank you very much

**Comment by pgrm** [comment]
ok, I've found the reason why `.FromSql` works with consts - the parameters are annotated with `[NotParameterized]`. Now everything works for me. I suppose you can close this issue?

Thanks

**Comment by MyEidos** [comment]
Could someone give example of  this service provider? I need to override some linq function and create SQL translation for it.

@pgrm, could you please share this part of you code, if it's possible.

**Comment by mikke49** [comment]
For anyone who tries to do this and stumbles over this article (like I did), you may find some more details and a solution example here: https://www.shadesofazure.com/sql-generation-ef-core/

It helped me when I finally found it :)

**Comment by TheThingRus** [comment]
@pgrm, @mikke49 could you please share example with others? :))
@mikke49, link doesn't work

**Comment by Systemko** [comment]
> @pgrm, @mikke49 could you please share example with others? :))
> @mikke49, link doesn't work

It seems that valid link is [https://www.chasingdevops.com/sql-generation-ef-core/](https://www.chasingdevops.com/sql-generation-ef-core/).
